cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -0,0 +1,732 @@
+                }
+                /* In case the packet is an ACK and the ACK filter is disabled,
+                 * indicate the frame reception */
+                else if (is_ack && !cfg.ack_filter) {
+                    dev->cb(dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+                }
+                /* If all failed, simply drop the frame and continue listening
+                 * to incoming frames */
+                else {
+                    NRF_RADIO->TASKS_START = 1;
+                }
+            }
+            else {
+                NRF_RADIO->TASKS_START = 1;
+            }
+            break;
+        case STATE_ACK:
+            _state = STATE_RX;
+            NRF_RADIO->PACKETPTR = (uint32_t) rxbuf;
+            _disable();
+            /* This will take around 0.5 us */
+            while (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {};
+            NRF_RADIO->TASKS_RXEN = 1;
+            _set_ifs_timer(false);
+            break;
+        default:
+            assert(false);
+        }
+    }
+
+    cortexm_isr_end();
+}
+
+static int _confirm_on(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    return 0;
+}
+
+static int _request_on(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    _state = STATE_IDLE;
+    NRF_RADIO->POWER = 1;
+    /* make sure the radio is disabled/stopped */
+    _disable();
+    /* we configure it to run in IEEE802.15.4 mode */
+    NRF_RADIO->MODE = RADIO_MODE_MODE_Ieee802154_250Kbit;
+    /* and set some fitting configuration */
+    NRF_RADIO->PCNF0 = ((8 << RADIO_PCNF0_LFLEN_Pos) |
+                        (RADIO_PCNF0_PLEN_32bitZero << RADIO_PCNF0_PLEN_Pos) |
+                        (RADIO_PCNF0_CRCINC_Include << RADIO_PCNF0_CRCINC_Pos));
+    NRF_RADIO->PCNF1 = IEEE802154_FRAME_LEN_MAX;
+    /* set start frame delimiter */
+    NRF_RADIO->SFD = IEEE802154_SFD;
+    /* set MHR filters */
+    NRF_RADIO->MHRMATCHCONF = 0;              /* Search Pattern Configuration */
+    NRF_RADIO->MHRMATCHMAS = 0xff0007ff;      /* Pattern mask */
+    /* configure CRC conform to IEEE802154 */
+    NRF_RADIO->CRCCNF = ((RADIO_CRCCNF_LEN_Two << RADIO_CRCCNF_LEN_Pos) |
+                         (RADIO_CRCCNF_SKIPADDR_Ieee802154 << RADIO_CRCCNF_SKIPADDR_Pos));
+    NRF_RADIO->CRCPOLY = 0x011021;
+    NRF_RADIO->CRCINIT = 0;
+
+    /* Disable the hardware IFS handling  */
+    NRF_RADIO->MODECNF0 |= RADIO_MODECNF0_RU_Msk;
+
+    NRF_RADIO->SHORTS = DEFAULT_SHORTS;
+
+    /* enable interrupts */
+    NVIC_EnableIRQ(RADIO_IRQn);
+    NRF_RADIO->INTENSET = RADIO_INTENSET_END_Msk |
+                          RADIO_INTENSET_CCAIDLE_Msk |
+                          RADIO_INTENSET_CCABUSY_Msk;
+
+    return 0;
+}
+
+static int _config_phy(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf)
+{
+    (void) dev;
+    _disable();
+    int8_t pow = conf->pow;
+
+    if (pow < TX_POWER_MIN || pow > TX_POWER_MAX) {
+        return -EINVAL;
+    }
+
+    NRF_RADIO->FREQUENCY = (((uint8_t) conf->channel) - 10) * 5;
```

Comment:
Is `5` channel spacing?

Reply:
I don't know, I simply copied this line from the original `netdev` driver.

Reply:
If it turns out that most radios will allow to set the frequency directly, there would be even more reason to move the channel calculations to the submac :wink: 
That would also ease supporting multiple regions in the sub-GHz band (868 MHz vs 902 MHz, etc)

Reply:
yes, we can do that. Actually, most O-QPSK radios already have a frequency register (even the `at86rf2xx`)

Reply:
I checked again. On `at86rf2xx` and `at86rf215` you set the base frequency (and channel spacing), then you can write the channel register and the radio will calculate frequency + chan * spacing internally.